alibi/explainers/anchor_image.py

```diff
@@ -613,7 +621,19 @@ def overlay_mask(self, image: np.ndarray, segments: np.ndarray, mask_features: l
     def _transform_predictor(self, predictor: Callable) -> Callable:
         # check if predictor returns predicted class or prediction probabilities for each class
         # if needed adjust predictor so it returns the predicted class
-        if np.argmax(predictor(np.zeros((1,) + self.image_shape, dtype=self.dtype)).shape) == 0:
+        x = np.zeros((1,) + self.image_shape, dtype=self.dtype)
+        try:
+            prediction = predictor(x)
+        except Exception as e:
```

Comment:
another point here is that in mlserver / alibi explainer server v1 we are wrapping a `[predict_fn`](https://github.com/SeldonIO/MLServer/blob/07a36f2cacf4eea6ed3ac28436792f31adb04d01/runtimes/alibi-explain/mlserver_alibi_explain/explainers/black_box_runtime.py#L57) around a REST call. So there might be other exceptions that are raised with the connection. I am not sure though how we will identify the differences here. 

interesting in the mlserver runtime it will raise this [exception](https://github.com/SeldonIO/MLServer/blob/07a36f2cacf4eea6ed3ac28436792f31adb04d01/runtimes/alibi-explain/mlserver_alibi_explain/common.py#L50).

not sure about the interop between libraries and how to deal with exceptions in this case.

Reply:
Good point, there must be a way around this... I guess the issue is that with these semantics `alibi` doesn't know what can go wrong calling a user model, so all exceptions are caught and re-raised with an error that assumes something about the model is misspecified. But for clients such as `mlserver` they would like to distinguish between these types of errors and network errors.

Reply:
Can we test the behaviour of this? I believe it could still work as is as the remote request shouldn't raise a network error but rather an error code, in which case if an error is raised it must have come from the model internals.

Reply:
I am raising `ValueError` and here is what we get.

```
args = ()
kwargs = {'predictor_url': 'dummy_call', 'v2_payload': InferenceRequest(id=None, parameters=Parameters(content_type='np'), inpu... 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]))], outputs=None)}
loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
res = InferenceResponse(model_name='custom_tf_mnist_model', model_version=None, id=None, parameters=None, outputs=[ResponseO...274435043, 0.10133041441440582, 0.09029622375965118, 0.11116134375333786, 0.07727331668138504, 0.09395473450422287]))])

    def mock_predict(*args, **kwargs):
        # note: sometimes the event loop is not running and in this case
        # we create a new one otherwise
        # we use the existing one.
        # this mock implementation is required as we dont want to spin up a server,
        # we just use MLModel.predict
        try:
            loop = asyncio.get_event_loop()
            res = loop.run_until_complete(
                custom_runtime_tf.predict(kwargs["v2_payload"])
            )
            raise ValueError("break")
            return res
        except Exception:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            res = loop.run_until_complete(
                custom_runtime_tf.predict(kwargs["v2_payload"])
            )
>           raise ValueError("break")
E           ValueError: break

conftest.py:161: ValueError

The above exception was the direct cause of the following exception:

args = ()
kwargs = {'anchor_image_directory': PosixPath('/home/sa/code/MLServer/runtimes/alibi-explain/tests/data/mnist_anchor_image'), 'custom_runtime_tf': <helpers.tf_model.TFMNISTModel object at 0x7f6936454f50>}
loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
setup = <function pytest_fixture_setup.<locals>.wrapper.<locals>.setup at 0x7f6a84016830>

    def wrapper(*args, **kwargs):
        loop = fixture_stripper.get_and_strip_from(
            FixtureStripper.EVENT_LOOP, kwargs
        )
    
        async def setup():
            res = await coro(*args, **kwargs)
            return res
    
>       return loop.run_until_complete(setup())

/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/pytest_asyncio/plugin.py:160: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/nest_asyncio.py:70: in run_until_complete
    return f.result()
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/asyncio/futures.py:181: in result
    raise self._exception
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/asyncio/tasks.py:249: in __step
    result = coro.send(None)
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/pytest_asyncio/plugin.py:157: in setup
    res = await coro(*args, **kwargs)
conftest.py:177: in anchor_image_runtime_with_remote_predict_patch
    await rt.load()
../mlserver_alibi_explain/runtime.py:165: in load
    return await self._rt.load()
../mlserver_alibi_explain/explainers/black_box_runtime.py:43: in load
    self._model = self._load_from_uri(self._infer_impl)
../mlserver_alibi_explain/runtime.py:91: in _load_from_uri
    return load_explainer(uri, predictor=predictor)
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/alibi/saving.py:68: in load_explainer
    return load_fn(path, predictor, meta)
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/alibi/saving.py:155: in _load_AnchorImage
    explainer.reset_predictor(predictor)
/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/alibi/explainers/anchor_image.py:643: in reset_predictor
    self.predictor = self._transform_predictor(predictor)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = AnchorImage(meta={
  'name': 'AnchorImage',
  'type': ['blackbox'],
  'explanations': ['local'],
  'params': {
       ...           'images_background': None,
              'segmentation_fn': 'slic'}
            ,
  'version': '0.6.2dev'}
)
predictor = <bound method AlibiExplainBlackBoxRuntime._infer_impl of <mlserver_alibi_explain.explainers.black_box_runtime.AlibiExplainBlackBoxRuntime object at 0x7f6935b20ad0>>

    def _transform_predictor(self, predictor: Callable) -> Callable:
        # check if predictor returns predicted class or prediction probabilities for each class
        # if needed adjust predictor so it returns the predicted class
        x = np.zeros((1,) + self.image_shape, dtype=self.dtype)
        try:
            prediction = predictor(x)
        except Exception as e:
            msg = f"Predictor failed to be called on {type(x)} of shape {x.shape} and dtype {x.dtype}. " \
                  f"Check that the parameter `image_shape` is correctly specified."
>           raise AlibiPredictorCallException(msg) from e
E           alibi.exceptions.AlibiPredictorCallException: Predictor failed to be called on <class 'numpy.ndarray'> of shape (1, 28, 28, 1) and dtype float32. Check that the parameter `image_shape` is correctly specified.

/home/sa/miniconda3/envs/python3.7-mlserver/lib/python3.7/site-packages/alibi/explainers/anchor_image.py:630: AlibiPredictorCallException
```

Reply:
@sakoush unless I'm misunderstanding, the above result looks good? As you are successfully raising the new `alibi` exception?

Reply:
I believe the point is that `mlserver` would like to catch specific errors e.g. `ValueError` from the scope calling the predictor, but because inside `alibi` all `Exception` are caught and re-raised with `AlibiPredictorCallException` this is not possible? Would be good to clarify again what the goals here are.

Reply:
@sakoush wouldn't this be solved by doing `except ValueError` instead? `mlserver` shouldn't really be catching bare exceptions and if has error handling for both `ValueError` and `AlibiPredictorCallException` we should be good to go.

Reply:
We are not actually capturing exceptions, what we do in `mlserver` when we get an error from calling the endpoint for `predict_fn` is 
```
    if response_raw.status_code != 200:
        raise RemoteInferenceError(response_raw.status_code, response_raw.reason)
```
we probably can be more specific for the error codes but I dont think that the statement `alibi.exceptions.AlibiPredictorCallException: Predictor failed to be called on <class 'numpy.ndarray'> of shape (1, 28, 28, 1) and dtype float32. Check that the parameter `image_shape` is correctly specified.` is valid generally. I need to think more about this from my side, basically how we can propagate exceptions interprocess?